### PR TITLE
Limit the use of `useTableColumnSort` hook

### LINKED ISSUE #1252: [Technical]: `useTableColumnSort` needs to not be publicly used by components
### Feature description

We have a Table component now that bridges our interface with that of PF components. 

However, we are still using `useTableColumnSort` [outside of it](https://github.com/search?q=repo%3Aopendatahub-io%2Fodh-dashboard++useTableColumnSort%3C+-path%3Afrontend%2Fsrc%2Fcomponents%2Ftable&type=code). That should no longer be the case -- the table already sorts the data. We should fix this.

Remove all references to the table sort outside of the table components/hooks -- ideally we even take the type out of the hook file and place it in a types file so it can be more generically referenced.

Ideally I'd like to see an index file at the root of `src/components/table` that exposes items that are intended for consumption and excludes the `useTableColumnSort` hook proper from that file. This way it drives home more that you're digging into implementation you shouldn't be doing. We might get a linter for this in the long run.

### Describe alternatives you've considered

_No response_

### Anything else?

Related to #1250

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
Closes #1252 

## Description

This PR did the following things:
1. Get rid of all the usage of the `useTableColumnSort` hook outside of the `~/components/table` repo
2. Create an `index.ts` file and export all the hooks, components, const, types out of the `~/components/table` repo
3. Create lint rules to limit only allow to import from `~/components/table` and don't allow to import from `~/components/table/**`. Also, limit the import of `useTableColumnSort`

## How Has This Been Tested?

These are mostly infrastructure refactors which will not bring any changes to the features and functionalities. The only thing you could test is to check whether the sort is still working on the Notebook Controller admin panel table and the data science projects table

## Test Impact

N/A, code refactor.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/components/table/index.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export * from './types';
+export * from './Table';
```
Comment (andrewballantyne) [inline frontend/src/components/table/index.ts:2]: I think this should be:
```suggestion
export { default as Table } from './Table'; 
```

Personally I don't think we need to remove the default exports for each file.

File: frontend/src/components/table/useCheckboxTable.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
 import { xor } from 'lodash';
-import Table from '~/components/table/Table';
+import { Table } from '~/components/table';
```
Comment (andrewballantyne) [inline frontend/src/components/table/useCheckboxTable.ts:3]: Careful with this -- you should import directly from the file in question and not use the index import for those items inside the index import. Circle references can occur in JavaScript Modules. I think there is decent protection these days and TypeScript doesn't suffer from this but it seems weird on face value to "have an index that is also imported by those inside the index"

Comment (andrewballantyne) [reply frontend/src/components/table/useCheckboxTable.ts:3]: An alternative would be to import just the type -- I don't see a need to import the component itself (ie, only do it for the TS compiler) -- tree shaking might prevent this in the webpack bundle 🤔 

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

Comment (DaoDaoNoCode) [reply frontend/src/components/table/useCheckboxTable.ts:3]: Good call, tried to use `import type Table from './Table'` and it works pretty good.

**Comment by DaoDaoNoCode** [comment]
/unhold
Code freeze ended.

File: frontend/.eslintrc [hunk 3]
```diff
@@ -63,6 +63,14 @@
...
             "group": ["~/api/**"],
             "message": "Read from '~/api' instead."
+          },
+          {
+            "group": ["~/components/table/**", "!~/components/table/useTableColumnSort"],
+            "message": "Read from '~/components/table' instead."
+          },
+          {
+            "group": ["~/components/table/useTableColumnSort"],
+            "message": "The data will be sorted in the table, don't use this hook outside of '~/components/table' repo. For more information, please check the props of the Table component."
```
Comment (christianvogt) [inline frontend/.eslintrc:73]: I think we should consider using a barrel file for our components API or adopt an `internal` folder structure hierarchy.
Not part of this PR, but for future tech debt. cc @andrewballantyne

Comment (christianvogt) [reply frontend/.eslintrc:73]: @andrewballantyne pointed me to this https://github.com/openshift/hac-infra#folder-structure-design which he suggests we adopt and I agree. I will make a separate tech debt item to capture this.

Comment (lucferbux) [reply frontend/.eslintrc:73]: https://github.com/opendatahub-io/odh-dashboard/issues/1833

**Review by Gkrumbach07 — APPROVED** [review]
/lgtm

**Review by andrewballantyne — APPROVED** [review]
Love the move to barrel files. It's a little verbose when an error happens, but the solution is in the lint error. So we can go with this for now, and work on improving it for later.

**Pull request merged**